### DISCUSSION TITLE
simple ipxe.iso edit

### DISCUSSION THREAD

**Discussion by ffgomez** [ipxe/ipxe#1155] [Q&A] [answered]
Dear all,

We are using the ipxe.iso by dding to the boot disk in remote machines as universal method (uefi, legacy, vms, baremetal,...) to force reboot in pxe mode. We also managed to select and configure the interface from the mac and launch RH/Centos installations, including the ks file (don't forget to wipefs!). Thus, we don't even depend on DHCP, which is also great for our environment.

However make-ing the iso per embeded script, therefore per machine, is quite a burden and also needing to install make, glibc, genisofs, ... in the controller node. (The plan is to ansibilize this for air-gapped environments)

Is there any possibility to dd an standard ipxe image (iso, hd, usb, ...) to the disk, mount that partition and then add ipxe-embed-script??

We tried many approachs to achieve this, but they failed so miserably that  I will save you the details.

Best regards and thanks in advance

**Comment by mcb30** [comment, marked as answer]
Yes, you can use `util/genfsimg` manually as follows:

```
make bin/ipxe.lkrn bin-x86_64-efi/ipxe.efi
./util/genfsimg -o ipxe.img -s embed.ipxe bin/ipxe.lkrn bin-x86_64-efi/ipxe.efi
```

This will create a combined BIOS/UEFI disk image which includes your `autoexec.ipxe` script.  You can subsequently edit the `autoexec.ipxe` script within the disk image.

You can also include an ARM64 image if you want a combined x86/ARM disk image, e.g.:
```
make bin/ipxe.lkrn bin-x86_64-efi/ipxe.efi
make bin-arm64-efi/ipxe.efi CROSS=aarch64-linux-gnu-
./util/genfsimg -o ipxe.img -s autoexec.ipxe bin/ipxe.lkrn bin-x86_64-efi/ipxe.efi bin-arm64-efi/ipxe.efi
```

**Reply by ffgomez** [reply]
Many thanks!

This will greatly alleviate the problem.

However, in addition to saving the genisoimage package, is there any possibility of creating a partition on the remote node (using dd, fdisk, mkfs), marking it as bootable (for legacy systems), and creating the /efi folder along with other necessary .lkrn files? I guess it won't be as universal as the ISO...

Best regards, and thanks again in advance!

**Reply by NiKiZe** [reply]
The suggestion is to create a disk image, write that to disk.
This is a FAT fs on a partition, which is the same as the ISO.

**Reply by mcb30** [reply]
@ffgomez You can't just create a generic partition with content and then add content to convert it into an iPXE boot image.  There are several constraints on how the disk image is laid out by `genfsimg` in order to create a hybrid BIOS/UEFI bootable image.

Use the disk image created as I describe above, write that to the disk, then edit it.

**Comment by ffgomez** [comment]
Thanks a lot for the answers!

I wasn't able to mount the usb/hd images (they are shown ans BeOS??) to add more content and the iso image is RO by definition. So I was confused about the procedure, sorry about that.

However by creating the image like:
 ./util/genfsimg -s ipxe_mac.ipxe -p 4000 -o ipxe-big.img bin/ipxe.lkrn bin-x86_64-efi/ipxe.efi
I was able to easily mount it and add more content. 

But the image wasn't working, I can see that it was able to load the lkrn and the script but then I got:
 "booting kernel failed: Invalid argument"

With my limited knowledge I compared the content of the .iso and .img and they were different:
.iso:
```
boot.catalog
ipxe.lkrn
isolinux.bin
isolinux.cfg
ldlinux.c32
```

.img:
```
autoexec.ipxe
EFI
ipxe.lkrn
ldlinux.c32
ldlinux.sys
syslinux.cfg
```

So I added the missing isolinux, boot.catalog and it is working. I read about the problem of not including isolinux and can be due a missing folder in my Redhat8.8 for isolinux, but that souldn't be the case as the ipxe.iso is also made from the source.

Would you be please so kind as to enlightme in this matter? I can't really get a grasp of what is going on.

Thank you so much again.

Best regards,

Felix

**Comment by mcb30** [comment]
> I wasn't able to mount the usb/hd images (they are shown ans BeOS??)

This is as expected: the raw `.usb` images do not contain any filesystem or syslinux code, just a small binary stub to load the raw binary directly.

> However by creating the image like: ./util/genfsimg -s ipxe_mac.ipxe -p 4000 -o ipxe-big.img bin/ipxe.lkrn bin-x86_64-efi/ipxe.efi I was able to easily mount it and add more content.
> 
> But the image wasn't working, I can see that it was able to load the lkrn and the script but then I got: "booting kernel failed: Invalid argument"

Did the image work **before** you mounted it and made changes?

> With my limited knowledge I compared the content of the .iso and .img and they were different: .iso:
> 
> ```
> boot.catalog
> ipxe.lkrn
> isolinux.bin
> isolinux.cfg
> ldlinux.c32
> ```
> 
> .img:
> 
> ```
> autoexec.ipxe
> EFI
> ipxe.lkrn
> ldlinux.c32
> ldlinux.sys
> syslinux.cfg
> ```

This is as expected: the .img format is not an ISO9660 filesystem and does not contain isolinux.

> Would you be please so kind as to enlightme in this matter? I can't really get a grasp of what is going on.

Start with the simplest test you can do, which is probably something like:

```
make bin/ipxe.lkrn
./util/genfsimg -o test.img bin/ipxe.lkrn
```

i.e. with no autoexec script, no extra padding, legacy BIOS only (no UEFI).  Test this image first without making any changes to it.  Then try gradually varying the test to add the bits you need one at a time.  When you encounter an error, post the full error text (including the surrounding context), or a screenshot if that's easier.

**Reply by ffgomez** [reply]
Thank you again.

It is now working flawlessly with all options. This time I tried to be methodical, testing every single option in isolation, and all of them worked, which I swear I did before with different results.

Now I noticed some problems with the script, but the first thing I tried was to test it without the script. I think I am facing a glitch in the Matrix.

Thanks for the awesome software, support, and patience!
